### DISCUSSION TITLE
Focusing on a button when I am on an element that contains it

### DISCUSSION THREAD

**Discussion by jakubziebin** [Textualize/textual#3351] [Q&A] [answered]
I have an element that contains some Static elements and one Button, structured like this:: 
```
    def compose(self) -> ComposeResult:
        with Horizontal():
            yield Static("Example static")
            yield Button("Example button", id="example-button")
```
Is there a way to achieve an effect similar to that of a classic button like this ?
![image](https://github.com/Textualize/textual/assets/116113682/cff76b3e-b12f-43d4-a315-6dd26be0f6db)

**Comment by davep** [comment]
Are you saying you want to make the `Horizontal` look like a `Button`, while also containing a `Static` and a `Button`?

**Reply by jakubziebin** [reply]
Not really, I only want this effect on the `Button` when I am on that element. Now, when I move through the app using `Tab`, I can't see which element I'm on.

**Reply by davep** [reply]
So is your question "How do I style a button to look like this?" -- if so you want to check [the documentation for `Button` where it documents the variants](https://textual.textualize.io/widgets/button/#example).

**Reply by jakubziebin** [reply]
Ok, thanks. But how can I go through the screen and when I focus on one of those `Horizontal` elements still have that button focus effect?

**Reply by jakubziebin** [reply]
So generally i want to have an effect: I am on `Horizontal` element -> i can see that i-am on this element by button.

**Reply by davep** [reply]
I wonder if there is something being left out here? By default `Horizontal` isn't an element that can receive focus. Do you think you could try rewording your question, perhaps with an illustrative example?

**Reply by jakubziebin** [reply]
Yes, so:
This is what the whole element looks like:
```
class Example(Widget):
    def compose(self) -> ComposeResult:
        yield Static("example_1", classes=odd)
        yield Static("example_2", classes=even)
        yield Button("example_3", id="example-button")
```
I have several such elements on the screen. When I go through the screen using `Tab` - I can't see which element I'm on and I want to have focus on `Button` when I'm on one of the `Example` element.
So - i am on `Example` and I want to have that focus effect on the button.
![image](https://github.com/Textualize/textual/assets/116113682/901ca817-487a-429e-88d5-3b30124382f1)

**Reply by davep** [reply]
What I'm struggling to understand about your question is this: only widgets that are capable of receiving focus can be focused; so given what you've shown above so far, as you <kbd>Tab</kbd> through your screen, you're only ever going to be moving from one `Button` to another anyway, so you should *always* see which element you're on. 

In the code you've shown here, `Example` can never receive focus, only the `Buttons` within each `Example`.

I think this is why it would be helpful here to see a fully-working minimal version of the problem you're dealing with.

**Reply by jakubziebin** [reply]
Yes, but i do not see that `Buttons` have focus - thats the problem. I`ll send example

**Reply by TomJGooding** [reply]
> When I go through the screen using `Tab` [...] I want to have focus on `Button` when I'm on one of the `Example` element.

Unless I am also struggling to understand what you mean, isn't this what Textual does already...?

I've quickly tried to recreate your example above (but based on Horizontal rather than Widget). When I click Tab, it focusses the next `Button`.

If this isn't what you're seeing, could you tell us what Textual version you're using?

```python
from textual.app import App, ComposeResult
from textual.containers import Horizontal
from textual.widgets import Button, Static


class Example(Horizontal):
    DEFAULT_CSS = """

    Example {
        height: 3;
        margin-bottom: 1;
    }

    .odd {
        background: $primary-background;
        width: 16;
        height: 3;
        content-align: center middle;
    }

    .even {
        background: $primary-background-darken-2;
        width: 16;
        height: 3;
        content-align: center middle;
    }
    """

    def compose(self) -> ComposeResult:
        yield Static("example_1", classes="odd")
        yield Static("example_2", classes="even")
        yield Button("example_3", id="example-button")


class MyApp(App):
    def compose(self) -> ComposeResult:
        for _ in range(5):
            yield Example()


if __name__ == "__main__":
    app = MyApp()
    app.run()
```

**Reply by jakubziebin** [reply]
Hm yes, you are right. I tried it on the latest version of textual and it is ok, but I am using `0.31.0` in the app. This is more or less the structure of my screen. 
```
from textual.app import App, ComposeResult
from textual.containers import Horizontal, ScrollableContainer
from textual.widgets import Button, Static, TabPane, TabbedContent


class Example(Horizontal):
    DEFAULT_CSS = """

    Example {
        height: 3;
        margin-bottom: 1;
    }

    .odd {
        background: $primary-background;
        width: 16;
        height: 3;
        content-align: center middle;
    }

    .even {
        background: $primary-background-darken-2;
        width: 16;
        height: 3;
        content-align: center middle;
    }
    """

    def compose(self) -> ComposeResult:
        yield Static("example_1", classes="odd")
        yield Static("example_2", classes="even")
        yield Button("example_3", id="example-button")


class NextExample(ScrollableContainer):
    def compose(self) -> ComposeResult:
        for _ in range(5):
            yield Example()


class ExampleTabPane(TabPane):
    def compose(self) -> ComposeResult:
        yield NextExample()


class MyApp(App):
    def compose(self) -> ComposeResult:
        with TabbedContent():
            yield ExampleTabPane("example")


if __name__ == "__main__":
    app = MyApp()
    app.run()
```

**Reply by davep** [reply, marked as answer]
Off the top of my head, I don't recall any changes after 0.31.0 that would change this behaviour; but if the latest version of Textual is working for you it would make sense to use the later version.

**Reply by TomJGooding** [reply]
> This is more or less the structure of my screen...

I've tested your example with the latest version of Textual and tabbing the Buttons works as expected.

After installing v0.31.0 and adding `height: auto` to the ScrollableContainer, the tabbing still works as expected.

So did you actually check your example demonstrates the issue you're describing? Or is there still something we are missing here?

**Reply by jakubziebin** [reply]
Yes, but I have solved the problem. I replaced the use of `Static` by `Label` and now it works fine.

**Reply by davep** [reply]
I very much suspect that something else will have affected the behaviour there; changing from `Static` to `Label` in the code you gave would make no difference to but focus works (moreover, `Label` is really just a `Static` with a different type, they're almost identical).
